torchmdnet/neighbors/common.cuh

```diff
@@ -0,0 +1,214 @@
+__device__ void addAtomPairToList(PairListAccessor<scalar_t>& list, int i, int j,
+                                  scalar3<scalar_t> delta, scalar_t distance,
+                                  bool add_transpose) {
+  const int32_t i_pair = atomicAdd(&list.i_curr_pair[0], add_transpose ? 2 : 1);
+  // Neighbors after the max number of pairs are ignored, although the pair is counted
+  if (i_pair + add_transpose < list.neighbors.size(1)) {
+    writeAtomPair(list, i, j, delta, distance, i_pair);
+    if (add_transpose) {
+      writeAtomPair(list, j, i, {-delta.x, -delta.y, -delta.z}, distance, i_pair + 1);
+    }
+  }
+}
+
+static void checkInput(const Tensor& positions, const Tensor& batch) {
+    TORCH_CHECK(positions.dim() == 2, "Expected \"positions\" to have two dimensions");
+    TORCH_CHECK(positions.size(0) > 0,
+                "Expected the 1nd dimension size of \"positions\" to be more than 0");
+    TORCH_CHECK(positions.size(1) == 3, "Expected the 2nd dimension size of \"positions\" to be 3");
+    TORCH_CHECK(positions.is_contiguous(), "Expected \"positions\" to be contiguous");
+
+    TORCH_CHECK(batch.dim() == 1, "Expected \"batch\" to have one dimension");
+    TORCH_CHECK(batch.size(0) == positions.size(0),
+                "Expected the 1st dimension size of \"batch\" to be the same as the 1st dimension "
+                "size of \"positions\"");
+    TORCH_CHECK(batch.is_contiguous(), "Expected \"batch\" to be contiguous");
+    TORCH_CHECK(batch.dtype() == torch::kInt64, "Expected \"batch\" to be of type torch::kLong");
+}
+
+namespace rect {
+
+/*
+ * @brief Takes a point to the unit cell in the range [-0.5, 0.5]*box_size using Minimum Image
+ * Convention
+ * @param p The point position
+ * @param box_size The box size
+ * @return The point in the unit cell
+ */
+template <typename scalar_t>
+__device__ auto apply_pbc(scalar3<scalar_t> p, scalar3<scalar_t> box_size) {
+    p.x = p.x - floorf(p.x / box_size.x + scalar_t(0.5)) * box_size.x;
+    p.y = p.y - floorf(p.y / box_size.y + scalar_t(0.5)) * box_size.y;
+    p.z = p.z - floorf(p.z / box_size.z + scalar_t(0.5)) * box_size.z;
+    return p;
+}
+
+template <typename scalar_t>
+__device__ auto compute_distance(scalar3<scalar_t> pos_i, scalar3<scalar_t> pos_j,
+                                 bool use_periodic, scalar3<scalar_t> box_size) {
+    scalar3<scalar_t> delta = {pos_i.x - pos_j.x, pos_i.y - pos_j.y, pos_i.z - pos_j.z};
+    if (use_periodic) {
+        delta = apply_pbc<scalar_t>(delta, box_size);
+    }
+    return delta;
+}
+
+} // namespace rect
+
+namespace triclinic {
+template <typename scalar_t> struct Box {
+    scalar_t size[3][3];
+    Box(const Tensor& box_vectors) {
+        if (box_vectors.size(0) == 3 && box_vectors.size(1) == 3) {
+            for (int i = 0; i < 3; i++) {
+                for (int j = 0; j < 3; j++) {
+                    size[i][j] = box_vectors[i][j].item<scalar_t>();
+                }
+            }
+        }
```

Comment:
I check that the box has the right size and that it lives in the CPU at the start of the operation.
I really need to know the box size CPU side, otherwise I cannot know how many cells are in each direction.
If I let the box live in the GPU this results in an unavoidable copy, so I force it to be in the CPU when constructing OptimizedDistance. Since it is just 9 numbers I can safely send it to kernels as an argument.
You are right about the if, though!